File: app/src/main/java/com/todoroo/astrid/adapter/CaldavTaskAdapter.java
package com.todoroo.astrid.adapter;

import static com.todoroo.andlib.utility.DateUtilities.now;

import com.todoroo.astrid.dao.TaskDao;
import com.todoroo.astrid.data.Task;
import org.tasks.data.CaldavDao;
import org.tasks.data.CaldavTask;
import org.tasks.data.TaskContainer;
import org.tasks.tasklist.ViewHolder;
import timber.log.Timber;

public final class CaldavTaskAdapter extends TaskAdapter {

  private final TaskDao taskDao;
  private final CaldavDao caldavDao;

  CaldavTaskAdapter(TaskDao taskDao, CaldavDao caldavDao) {
    this.taskDao = taskDao;
    this.caldavDao = caldavDao;
  }

  @Override
  public int getIndent(TaskContainer task) {
    return task.getIndent();
  }

  @Override
  public boolean canMove(ViewHolder sourceVh, ViewHolder targetVh) {
    TaskContainer source = sourceVh.task;
    int to = targetVh.getAdapterPosition();
    return !taskIsChild(source, to);
  }

  @Override
  public int maxIndent(int previousPosition, TaskContainer task) {
    TaskContainer previous = getTask(previousPosition);
    return previous.getIndent() + 1;
  }

  @Override
  public int minIndent(int nextPosition, TaskContainer task) {
    return 0;
  }

  @Override
  public boolean supportsParentingOrManualSort() {
    return true;
  }

  @Override
  public void moved(int from, int to, int indent) {
    TaskContainer task = getTask(from);
    TaskContainer previous = to > 0 ? getTask(to - 1) : null;

    long newParent = task.getParent();
    if (indent == 0) {
      newParent = 0;
    } else if (previous != null) {
      if (indent == previous.getIndent()) {
        newParent = previous.getParent();
      } else if (indent > previous.getIndent()) {
        newParent = previous.getId();
      }
    }

    // If nothing is changing, return
    if (newParent == task.getParent()) {
      return;
    }

    changeParent(task, newParent);

    Task update = task.getTask();
    update.setModificationDate(now());
    taskDao.save(update);
  }

  private void changeParent(TaskContainer task, long newParent) {
    CaldavTask caldavTask = task.getCaldavTask();

    if (newParent == 0) {
      caldavTask.setRemoteParent("");
      caldavTask.setParent(0);
    } else {
      CaldavTask parentTask = caldavDao.getTask(newParent);
      if (parentTask == null) {
        return;
      }
      caldavTask.setRemoteParent(parentTask.getRemoteId());
      caldavTask.setParent(newParent);
    }
    caldavDao.update(caldavTask);
  }

  private boolean taskIsChild(TaskContainer source, int destinationIndex) {
    TaskContainer destination = getTask(destinationIndex);
    while (destination.getParent() != 0) {
      if (destination.getParent() == source.getParent()) {
        return false;
      }
      if (destination.getParent() == source.getId()) {
        return true;
      }
      destination = getTask(--destinationIndex);
    }
    return false;
  }
}

File: app/src/main/java/com/todoroo/astrid/adapter/CaldavTaskAdapter.kt
package com.todoroo.astrid.adapter

import com.todoroo.astrid.dao.TaskDao
import org.tasks.data.CaldavDao
import org.tasks.data.TaskContainer
import org.tasks.date.DateTimeUtils.toAppleEpoch

open class CaldavTaskAdapter internal constructor(private val taskDao: TaskDao, private val caldavDao: CaldavDao, private val newTasksOnTop: Boolean = false) : TaskAdapter() {
    override fun supportsParentingOrManualSort() = true

    override fun moved(from: Int, to: Int, indent: Int) {
        val task = getTask(from)
        val newParent = changeParent(task, indent, to)
        if (newParent != task.parent) {
            val newPosition = if (newTasksOnTop) {
                caldavDao.findFirstTask(task.caldav!!, newParent)
                        ?.takeIf { task.creationDate.toAppleEpoch() >= it}
                        ?.minus(1)
            } else {
                caldavDao.findLastTask(task.caldav!!, newParent)
                        ?.takeIf { task.creationDate.toAppleEpoch() <= it }
                        ?.plus(1)
            }
            caldavDao.update(task.caldavTask.cd_id, newPosition)
        }
    }

    internal fun changeParent(task: TaskContainer, indent: Int, to: Int): Long {
        val newParent = findParent(indent, to)?.id ?: 0
        if (task.parent != newParent) {
            changeParent(task, newParent)
        }
        return newParent
    }

    private fun changeParent(task: TaskContainer, newParent: Long) {
        val caldavTask = task.getCaldavTask()
        if (newParent == 0L) {
            caldavTask.cd_remote_parent = ""
            task.parent = 0
        } else {
            val parentTask = caldavDao.getTask(newParent) ?: return
            caldavTask.cd_remote_parent = parentTask.remoteId
            task.parent = newParent
        }
        caldavDao.update(caldavTask.cd_id, caldavTask.cd_remote_parent)
        taskDao.save(task.getTask(), null)
    }
}